### PR TITLE
Avoid index out of range issue

### PR DESCRIPTION
**Pull request by lcs-rgordon**
When a lot of commands are added in a short period of time, and we are animating the drawing, TortoiseCharmer seems to get tripped up on the contents of the commands array of a Tortoise instance.

Specifically, the Tortoise.commands array seems to change size while TortoiseCharmer.charm is trying to do it's thing, which causes an index out of range error.

screen shot 2018-11-09 at 5 06 10 pm

The commit in this pull request contains the bare minimum code that I came up with to fix, or at least work around, the issue.

The issue can be reproduced by modifying the single tortoise playground like this:

screen shot 2018-11-09 at 5 24 56 pm

In this commit on another branch of my TortoiseGraphics fork, I spent time digging in to this:

https://github.com/lcs-rgordon/TortoiseGraphics/commit/8bbe04329813f54428d3e11bd52354cff2f00102

1. In that branch I've created a macOS app and used the PlaygroundCanvas class to view the animation so that the full debugger can be used to inspect the index out of range error.

2. I've included code that reliably triggers the index out of range error (similar to a recent assignment I gave my students).

### CODE REVIEW

**Comment by lcs-rgordon** [comment]
Additional notes...

While the code in this pull request fixes the issue (in my testing) I feel like there may be a better solution.

The index out of range issue only occurs when drawing is animated (using the PlaygroundCanvas class) and does not occur when GraphicsCanvas is used.

When PlaygroundCanvas animates the drawing, are there multiple threads operating on the the Tortoise.commands array?

**Comment by temoki** [comment]
@lcs-rgordon Thank you for your PR.

Perhaps the `TortoiseCharmer`'s `commandedHandler` property may be accessed from multiple threads. I think It should be locked while writing and reading the property. I added some codes to fix (I did not check whether it works or not). Could you try this code? 

```swift
class TortoiseCharmer {

    private(set) var tortoises: [Tortoise] = []
    private(set) var commandHistories: [(tortoiseTag: Int, commandIndex: Int)] = []
    
    // ⭐️ Add lock object.
    private let lock = NSLock()

    var commandedHandler: ((TortoiseCharmer) -> Void)?

    init(tortoiseCount: Int = 1) {
        initialize(tortoiseCount: tortoiseCount)
    }

    func initialize(tortoiseCount: Int = 1) {
        assert(tortoiseCount > 0)
        tortoises = []
        commandHistories = []
        for index in 0..<tortoiseCount {
            let tortoise = Tortoise()
            tortoise.tag = index
            tortoise.commandedHandler = { [unowned self] (tortoise) in
                // ⭐️ Lock while writing commandHistories.
                defer { self.lock.unlock() }
                self.lock.lock()
                self.commandHistories.append((tortoiseTag: tortoise.tag,
                                              commandIndex: tortoise.commands.count - 1))
                self.commandedHandler?(self)
            }
            tortoises.append(tortoise)
        }
    }

    @discardableResult
    func charm(with context: GraphicsContext, toFrame index: Int?) -> Int {
        // ⭐️ Lock while reading commandHistories.
        defer { self.lock.unlock() }
        self.lock.lock()

        let endIndex = commandHistories.count - 1
        let toIndex = min(max((index ?? endIndex), 0), endIndex)

        context.setup()

        var states = [State](repeating: State(), count: tortoises.count)
        for index in 0..<tortoises.count {
            states[index].canvasSize = context.size
        }

        for (index, history) in commandHistories.enumerated() where index <= toIndex {
            states[history.tortoiseTag] =
                tortoises[history.tortoiseTag]
                    .commands[history.commandIndex]
                    .exexute(in: states[history.tortoiseTag], with: context.cgContext)
        }

        for index in 0..<tortoises.count {
            let state = states[index]
            if state.isVisible {
                tortoises[index].drawTortoise(context.cgContext, state: state)
            }
        }

        context.tearDown()
        return min(toIndex + 1, endIndex)
    }

}

```

**Comment by lcs-rgordon** [comment]
👍🏻 Thanks! I will try this.

( I was looking at the code again just now. Definitely seeing `commandedHandler` touched on different threads. Will get back to you in a few minutes. It's been a long time since I've dealt with code running multiple threads. 😅 )

**Comment by temoki** [comment]
@lcs-rgordon Oops, I just tried it now, it does not works :cry:.

Perhaps the tortoise's commands property is written in other threads... 

```swift
        for (index, history) in commandHistories.enumerated() where index <= toIndex {
            states[history.tortoiseTag] =
                tortoises[history.tortoiseTag]
                    .commands[history.commandIndex] // ⭐️ 
                    .exexute(in: states[history.tortoiseTag], with: context.cgContext)
        }
```

**Comment by lcs-rgordon** [comment]
Hmm. The code you'd suggested seems to be working for me, but animation performance is noticeably impacted (working on a mid-2012 MacBook Pro, so not super speedy).

If there's a way to lock threads while not impacting animation speed, that feels like the safer and more correct fix.

If that's not possible, what do you think of the changes I've just committed to this PR?

https://github.com/temoki/TortoiseGraphics/pull/37/commits/d8757818ff3dc5f44c71ab73c4acdcfe98b9710e

**Comment by lcs-rgordon** [comment]
Completely unscientific, but here are two screen recordings showing the impact on animation speeds.

With the code that looks for commandHistories to stop changing length (i.e.: main thread is finished adding commands to be animated):

https://youtu.be/fKew88Y5MTQ

With the lock object (about 20 seconds longer to animate):

https://youtu.be/MPrDyhoisaE

**Comment by lcs-rgordon** [comment]
To reply to your earlier thought, yes, I think the main thread is adding commands to each `Tortoise` instance (I set a breakpoint and followed the code down to the `add` method in `Tortoise` class):

screen shot 2018-11-11 at 9 52 49 am

Immediately after that, the `commandedHandler` block is invoked.

In my testing the other day, it was accessing the Tortoise commands array that was causing the issue.

Can a lock be placed in the `add` method of the `Tortoise` class?

screen shot 2018-11-11 at 10 00 06 am

So that when a `Tortoise` instance on the main thread adds a command to its array,  the `TortoiseCharmer` instance running on the animation thread doesn't get to read the array until the other thread is finished modifying the list of commands?

This might impact performance less, as there is locking only when the `Tortoise` commands array is being populated (as opposed to locking every time a frame is drawn by `TortoiseCharmer`).

**Comment by temoki** [comment]
@lcs-rgordon 

> Can a lock be placed in the add method of the Tortoise class?

I tried your suggestion, but 'index out of range' still happens.

---

This commit looks not good to me.

https://github.com/lcs-rgordon/TortoiseGraphics/commit/d8757818ff3dc5f44c71ab73c4acdcfe98b9710e

Your first commit ( https://github.com/temoki/TortoiseGraphics/pull/37/commits/8f32ab34698f113988b3596950a3d184adefc246 ) is better, I think that it may be merged.

---

Do you want to animate smoothly? 
I just have an idea with CAShapeLayer.

**Comment by lcs-rgordon** [comment]
Sounds good to me!

I’m not super familiar with CAShapeLayer, but I’m happy to learn. 😀

**Comment by lcs-rgordon** [comment]
> In fact, like practically every Core Animation animation, once a shape layer animation is committed from the main thread of an application, Core Animation does all the heavy lifting inside its own processes, leaving an application’s main thread completely free during the animation.

and 

> The primary advantage of using shape layers for animation in place of using something like Core Graphics is that `CAShapeLayer` is composited on the GPU, making it significantly faster

from [CAShapeLayer in Depth](https://www.calayer.com/core-animation/2017/12/25/cashapelayer-in-depth-part-ii.html).

@temoki I do think I see where you are going with this 😀

**Comment by lcs-rgordon** [comment]
Operating quite a bit out of my comfort zone, I played around with animating between two paths [in a playground this afternooon](https://github.com/lcs-rgordon/TortoiseGraphics/blob/cashapelayer-experiment/Playground/Playground.playground/Pages/CAShapeLayer%20Test.xcplaygroundpage/Contents.swift). 

When it runs, it’s clear that the playground executes all statements well before the animation completes, which demonstrates the built-in threading provided by `CAShapeLayer`. 

It would be a major re-write, but could TortoiseGraphics be refactored to avoid the use of command arrays entirely? Just add to a path and animate using CAShapeLayer with a transition time in seconds of 1/framerate?

**Comment by lcs-rgordon** [comment]
[This](https://www.calayer.com/core-animation/2016/05/22/cashapelayer-in-depth.html#fn:5) might be a problem:

screen shot 2018-11-11 at 8 17 05 pm

It seems as if a given CAShapeLayer instance keeps global properties for all paths rendered within it.

So if I'm understanding this correctly, it would not be possible to change the pen size for the turtle from one line segment that is drawn to the next (or color for that matter).

I've tested this a bit using the playground I linked to above. I haven't been able to get paths to have different style attributes.

**Comment by lcs-rgordon** [comment]
Sorry. I know I've been posting a lot on this PR.

About performance, the only place I see issues now is at the end of a long animation. A student made the following (I adjusted only the name of one variable for clarity):

```swift
import PlaygroundSupport
import TortoiseGraphics
import CoreGraphics

let myFrame = CGRect(x: 0, y: 0, width: 600, height: 600)
let canvas = PlaygroundCanvas(frame: myFrame)
canvas.frameRate = 300
canvas.color = .black
PlaygroundPage.current.liveView = canvas

var size = 300
var colourToggle = 0

canvas.drawing { turtle in
    
    turtle.hideTortoise()
    
    turtle.fillColor(.black)
    turtle.penColor(.black)
    turtle.penSize(3)
    
    for _ in 1...300 {
        
        turtle.beginFill()
        for _ in 1...4 {
            turtle.forward(Double(size))
            turtle.right(90)
        }
        
        turtle.endFill()
        
        colourToggle += 1
        if colourToggle % 2 == 0 {
            turtle.fillColor(.black)
        } else {
            turtle.fillColor(.white)
        }
        
        size -= 1
        turtle.right(4)
        
    }
    
}

```

Exactly the kind of experimentation I try to encourage from students. Here's what is looks like when it's complete:

screen shot 2018-11-11 at 3 00 18 pm

With the revisions already committed in this PR, the code above no longer causes `Index out of range` errors.

However, by the end of the animation, rendering really starts to lag.

If I'm reading `TortoiseCharmer` and the other source code correctly, that's because for each frame in the animation, the graphics context begins as a blank slate, and then the drawing is built up from the very first frame to the current frame, and then the image is displayed.

I'm wondering if `PlaygroundCanvas` might be modified such that a single graphics context (an offscreen representation?) is used across frames. Each command adds to the drawing in that context, and when that command has been executed, the updated image (generated from the offscreen representation) is shown in the view?

I did something [sort of like that here](https://github.com/lcs-ics3u-2017/Animation/blob/master/CanvasTestArea.playground/Sources/Helper.swift) when I wrote a [Processing](https://processing.org)-like playground for use by my students, but I could never figure out how to get the playground's view to be animated like it is with `TortoiseGraphics`.

I am very far from an expert with any of this; have simply been motivated to teach using graphics in the style of Seymour Papert. I am thrilled that `TortoiseGraphics` exists.

I could try implementing the changes I've proposed above, but realistically the next time I'd have the opportunity is late December / early January, my next extended break from daily teaching duties. I've had the time to dive in to the `TortoiseGraphics` codebase over the last few days because school is out of session for a short November break.

@temoki whatever direction future development of `TortoiseGraphics` takes, I hope I can find ways to contribute and help out. Thank-you!

**Comment by temoki** [comment]
@lcs-rgordon Thanks a lot :smile:

Because I am not good at English, it takes time to read and understand your comment. I am very sleepy today so I will read it tomorrow. :bow:

**Comment by lcs-rgordon** [comment]
I understand. Thanks! Be well.

**Comment by temoki** [comment]
@lcs-rgordon 

https://github.com/temoki/TortoiseGraphics/pull/37#issuecomment-437710713

> When it runs, it’s clear that the playground executes all statements well before the animation completes, which demonstrates the built-in threading provided by CAShapeLayer.

You should set the `PlaygroundPage`'s `needsIndefiniteExecution` property.

https://developer.apple.com/documentation/playgroundsupport/playgroundpage/1964501-needsindefiniteexecution

**Comment by temoki** [comment]
https://github.com/temoki/TortoiseGraphics/pull/37#issuecomment-437725295

> So if I'm understanding this correctly, it would not be possible to change the pen size for the turtle from one line segment that is drawn to the next (or color for that matter).

I know. It is task to solve.

**Comment by lcs-rgordon** [comment]
@temoki Do you think [this](https://github.com/temoki/TortoiseGraphics/blob/8f32ab34698f113988b3596950a3d184adefc246/Sources/TortoiseCharmer.swift), the original suggestion for a fix to the "index out of range" issue could be merged?

The long-term improvement (using CAShapeLayer), or another approach, could be added as an issue.

Right now, until some kind of fix is merged, TortoiseGraphics isn't super-usable in a playground environment because the "index out of range" issue happens so frequently. It might be nice to fix that now and then look for a better fix as a separate item.

**Comment by temoki** [comment]
@lcs-rgordon I commented as follows on https://github.com/temoki/TortoiseGraphics/pull/37#issuecomment-437679414 .

> Your first commit ( 8f32ab3 ) is better, I think that it may be merged.

I'll merge this pull request if you revert to the first commit.

**Comment by lcs-rgordon** [comment]
Reverted as suggested. Thanks!

**Review by temoki — APPROVED** [review]

**Comment by temoki** [comment]
@lcs-rgordon LGTM 👍

**Pull request merged**
